File: templates/typescript-tests/src/views/dictionaryTests.ts
/**
 * OlapDictionary E2E test definitions.
 *
 * Defines a simple integer-keyed lookup dictionary over the existing
 * IndexTestTable to exercise the full dictionary lifecycle:
 * TypeScript definition → Rust parsing → migration DDL → ClickHouse.
 */
import { OlapDictionary, UInt64 } from "@514labs/moose-lib";
import { IndexTestTable } from "../ingest/models";

/**
 * Attribute shape for the IndexTest dictionary lookup.
 * All columns from IndexTest are available as dictionary attributes.
 */
interface IndexTestLookup {
  u64: UInt64; // primary key — UInt64 integer key for HASHED layout
  i32: number; // attribute
  s: string; // attribute
}

/**
 * A HASHED dictionary over IndexTestTable.
 * UInt64 primary key → HASHED layout (fastest for numeric keys).
 * lifetime: 3600 → reload from ClickHouse at most once per hour.
 */
export const indexTestLookupDict = new OlapDictionary<IndexTestLookup>(
  "dict_index_test_lookup",
  {
    sourceTable: IndexTestTable,
    primaryKey: ["u64"],
    layout: { type: "HASHED" },
    lifetime: 3600,
  },
);

File: templates/python-tests/src/views/dictionary_tests.py
"""
OlapDictionary E2E test definitions (Python mirror of dictionaryTests.ts).

Defines a simple integer-keyed lookup dictionary over the existing
index_test_table to exercise the full dictionary lifecycle:
Python definition → Rust parsing → migration DDL → ClickHouse.
"""

from moose_lib.dmv2 import (
    OlapDictionary,
    OlapDictionaryConfig,
    HashedLayout,
    DictionaryLifetime,
)
from src.ingest.models import index_test_table, IndexTest

# A HASHED dictionary over index_test_table.
# UInt64 primary key → HASHED layout (fastest for numeric keys).
# lifetime: static (min=0, max=0) → never auto-reloads.
index_test_lookup_dict = OlapDictionary[IndexTest](
    "dict_index_test_lookup",
    OlapDictionaryConfig(
        source_table=index_test_table,
        primary_key=["u64"],
        layout=HashedLayout(),
        lifetime=DictionaryLifetime(min=0, max=0),
    ),
)
